### PR TITLE
Branch/20 merge contacts

### PR DESCRIPTION
**Pull request by timohuber**

### CODE REVIEW

File: pool/app/duplicate_contacts/entity.ml [hunk 1]
```diff
@@ -0,0 +1,98 @@
+module Field = Pool_message.Field
+
+(* The script will determine a score between 0 and 1 based on how equal two
+   users are. The alert_threshold is the value from which it will be seen as a
+   possible duplicate *)
+
+let alert_threshold = 0.7
```
Comment (mabiede) [inline pool/app/duplicate_contacts/entity.ml:7]: Should this be customizable by tenant?

File: pool/app/duplicate_contacts/entity.ml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    ; sql_table : string
+    ; weight : int
+    }
+  [@@deriving eq, show]
+end
+
+let columns =
+  [ Field.Name, SimilarityCriteria.Exact, "user_users", "name", 5
+  ; Field.Firstname, SimilarityCriteria.Exact, "user_users", "given_name", 4
+  ; Field.CellPhone, SimilarityCriteria.Exact, "pool_contacts", "cell_phone", 3
```
Comment (mabiede) [inline pool/app/duplicate_contacts/entity.ml:38]: Shouldn't a cell phone have a higher weight than a "given name" or "name"?

File: pool/app/duplicate_contacts/repo.ml [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+;;
+
+let find_request_sql ?(count = false) =
+  let columns =
+    if count then "COUNT(*)" else sql_select_columns |> CCString.concat ", "
+  in
+  Format.asprintf
+    {sql| SELECT %s FROM pool_contacts_possible_duplicates %s %s |sql}
+    columns
+    joins
```
Comment (mabiede) [inline pool/app/duplicate_contacts/repo.ml:34]: used `[%string ...]` like the following SQL string concatenations

```suggestion
let find_request_sql ?(count = false) where =
  let columns =
    if count then "COUNT(*)" else sql_select_columns |> CCString.concat ", "
  in
  [%string
    {sql| SELECT %{columns} FROM pool_contacts_possible_duplicates %{joins} %{where} |sql}]
```

File: pool/app/duplicate_contacts/service.ml [hunk 4]
```diff
@@ -0,0 +1,48 @@
+let src = Logs.Src.create "user_import.service"
+let get_or_failwith = Pool_common.Utils.get_or_failwith
+let interval_s = 5
```
Comment (mabiede) [inline pool/app/duplicate_contacts/service.ml:3]: Do we need to run this every 5 seconds? IMO also every hour would be fine.

File: pool/app/pool_common/locales/i18n_en.ml [hunk 5]
```diff
@@ -345,6 +346,11 @@ When inviting contacts, the filter will prefer the overriding value if both are
     "Please take into account that the data quality is lower for text entries. If the \
      data can be collected in another form, this is preferable."
   | CustomHtmx s -> s
+  | DashboardDuplicateContactsNotification count ->
+    Format.asprintf
+      "%i possible diplicate %s have been found. Please take the necessary measures."
+      count
+      (if count = 1 then "contact" else "contacts")
```
Comment (mabiede) [inline pool/app/pool_common/locales/i18n_en.ml:353]:
```suggestion
      "%i possible duplicate %s been found. Please take the necessary measures."
      count
      (if count = 1 then "contact has" else "contacts have")
```

File: pool/app/pool_database/migrations/migration_202408081359.ml [hunk 6]
```diff
@@ -17,18 +17,28 @@ let create_changelog_table =
     |sql}
 ;;
 
+(* TODO: Fix manually on staging *)
```
Comment (mabiede) [inline pool/app/pool_database/migrations/migration_202408081359.ml:20]: Task for issue, not in code. Can be removed.

File: pool/command/command_migrate.ml [hunk 7]
```diff
@@ -48,7 +48,9 @@ let root =
     let%lwt () =
       Migration.execute Pool.Root.label (Pool_database.Root.steps ())
       ||> function
-      | Ok () -> exit 0
+      | Ok () ->
+        (* TODO: why is exit 0 not stopping the process? *)
```
Comment (mabiede) [inline pool/command/command_migrate.ml:52]: open ToDo

File: pool/cqrs_command/duplicate_contacts_command.ml [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    =
+    Logs.info ~src (fun m -> m "Handle command Merge" ~tags);
+    let open CCResult.Infix in
+    let open Pool_message in
+    let open Duplicate_contacts in
+    let select_contact_by_id id =
+      let id = Contact.Id.of_string id in
+      if Contact.Id.equal id (Contact.id contact_a) then contact_a else contact_b
+    in
+    let select_fields id =
```
Comment (mabiede) [inline pool/cqrs_command/duplicate_contacts_command.ml:38]: consistency

```suggestion
    let select_fields_by_id id =
```

File: pool/test/custom_field_utils.ml [hunk 9]
```diff
@@ -0,0 +1,606 @@
...
+          ; name
+          ; hint
+          ; validation
+          ; required
+          ; admin_override
+          ; admin_input_only
+          ; prompt_on_registration
+          ; version
+          }
+        , answer )
```
Comment (mabiede) [inline pool/test/custom_field_utils.ml:240]: predefine field before and reuse in switch

```suggestion
    let field ?(validation = Validation.pure) () =
      { Public.id
      ; name
      ; hint
      ; validation
      ; required
      ; admin_override
      ; admin_input_only
      ; prompt_on_registration
      ; version
      }
    in
    match field_type with
    | FieldType.Boolean ->
      let answer =
        { Answer.id = answer_id; entity_uuid; value = Some true; admin_value }
        |> CCOption.pure
      in
      Public.Boolean (field (), answer)
    | FieldType.Date ->
      let answer =
        { Answer.id = answer_id; entity_uuid; value = Some (1970, 1, 1); admin_value }
        |> CCOption.pure
      in
      Public.Date (field (), answer)
    | FieldType.MultiSelect ->
      let answer =
        field_options
        |> CCList.head_opt
        |> CCOption.map CCList.pure
        |> Answer.create ~id:answer_id entity_uuid
      in
      let validation = validation_schema Validation.MultiSelect.schema in
      Public.MultiSelect (field ~validation (), field_options, Some answer)
    | FieldType.Number ->
      let answer =
        { Answer.id = answer_id; entity_uuid; value = Some 3; admin_value }
        |> CCOption.pure
      in
      let validation = validation_schema Validation.Number.schema in
      Public.Number (field ~validation (), answer)
    | FieldType.Select ->
      let answer =
        CCList.head_opt field_options
        |> fun option ->
        { Answer.id = answer_id; entity_uuid; value = option; admin_value }
        |> CCOption.pure
      in
      Public.Select (field (), field_options, answer)
    | FieldType.Text ->
      let answer =
        { Answer.id = answer_id; entity_uuid; value = Some "test"; admin_value }
        |> CCOption.pure
      in
      let validation = validation_schema Validation.Text.schema in
      Public.Text (field ~validation (), answer)
```

**Review by mabiede — CHANGES_REQUESTED** [review]
- [ ] Database migration fails on stage: `add fk constraint to users`

**Comment by mabiede** [comment]
@timohuber 🚀 UI and selection of fields is very nice

File: pool/app/duplicate_contacts/service.ml (hunk 4 above)
Comment (timohuber) [reply pool/app/duplicate_contacts/service.ml:3]: Yes, the command only looks for duplicates of one contact.

File: pool/app/pool_database/migrations/migration_202408081359.ml (hunk 6 above)
Comment (timohuber) [reply pool/app/pool_database/migrations/migration_202408081359.ml:20]: Fixed

File: pool/app/duplicate_contacts/entity.ml (hunk 1 above)
Comment (timohuber) [reply pool/app/duplicate_contacts/entity.ml]: As discussed: We will see if this is necessary

File: pool/app/duplicate_contacts/entity.ml (hunk 2 above)
Comment (timohuber) [reply pool/app/duplicate_contacts/entity.ml]: As discussed: Adjusted

**Review by mabiede — APPROVED** [review]

**Pull request merged**